Common build/dev environment

ocefpaf opened the pull request:
Not ready yet. I am testing the build/dev env on the CI and locally to get a working env for the future changes.

Pending on https://github.com/conda-forge/staged-recipes/pull/2197

ocefpaf commented on .travis.yml:
```diff
@@ -19,19 +19,22 @@ matrix:
 
 before_install:
-    - pip install -r requirements.txt
-    - pip install -r requirements-dev.txt
```
@miguelcleon I am changing everything to conda so we can have more control over the packages.
Building a working env with `pip` should still be possible, but geodjango stuff is much easier with conda.

miguelcleon replied:
@ocefpaf  Ok I'll try it out. Yeah, I'm not surprised you ran into an issue with geodjango, I've run into a couple.

ocefpaf replied:
It depends on `gdal` and that is always painful to install. But I maintain `django` and `gdal` on conda-forge, so we at least have a man on the inside to request improvements :smile:

lsetiawan replied:
Woo! You're awesome Filipe! 👌

miguelcleon replied:
@ocefpaf one thing we can do is remove django-uuidfield. We aren't using that anymore. Django added this functionality into core django in version 1.9

ocefpaf replied:
Removing that...

miguelcleon replied:
hmm I don't see it in odm2adminenv.yml or requirements.txt. Yet I'm getting 

`UnsatisfiableError: The following specifications were found to be in conflict:
  - django-uuidfield -> django >=1.7,<1.8
  - django-uuidfield -> python 3.4*
  - python 2.7*
`

ocefpaf replied:
Yep, that is why I marked as "Not ready yet" :smile: 

I am working on the other side, `conda-forge`, to include the remaining packages.

miguelcleon replied:
cool, sounds good.

ocefpaf commented on .travis.yml:
```diff
@@ -19,19 +19,22 @@ matrix:
-    #- python setup.py sdist && version=$(python setup.py --version) && pushd dist  && pip install odm2djangoadmin-${version}.tar.gz && popd
+
+install:
+  - python setup.py sdist && version=$(python setup.py --version) && pushd dist  && pip install odm2djangoadmin-${version}.tar.gz && popd
```
Test the tarball creation.

ocefpaf commented on .travis.yml:
```diff
@@ -19,19 +19,22 @@ matrix:
-        find . -type f -name "*.py" ! -name "base.py" | xargs flake8 --max-line-length=100 --ignore=E402 ;
-      fi
+  - if [[ $TEST_TARGET == 'default' ]]; then
+      python setup.py test;
```
@lsetiawan I believe there are no unit nor integration tests yet, but this would catch anything we may write in the future. (I recommend writing with pytest btw :wink: )

lsetiawan replied:
@ocefpaf Hmm... interesting. I will look into unit and integration tests further and see if I can integrate it in.. I'm learning something new everyday! This is fun! 😺

ocefpaf replied:
That line calls  [this](https://github.com/miguelcleon/ODM2-Admin/blob/master/setup.py#L7-L15) code block btw.

lsetiawan replied:
Cool! Thanks! I was wondering what that function does.. Now I know. 😄

ocefpaf commented on odm2adminenv.yml:
```diff
@@ -1,17 +1,20 @@
-    - django-apptemplates
-    - django-daterange-filter
-    - django-import-export
-    - django-jquery
```
I packaged all but `django-recaptcha` in conda-forge. Still needs testing though...

ocefpaf commented on odm2adminenv.yml:
```diff
@@ -1,17 +1,19 @@
+  - django-import-export
+  - django-jquery
   - django=1.10
+  - djangocms-admin-style==0.2.7
```
@miguelcleon is this pin necessary?

(Just curious there is no need to remove it.)

lsetiawan replied:
@ocefpaf For some reason, if I don't use that exact version. The webpage doesn't render correctly... Maybe it's just on my end. Not sure what they changed from ver. 0.2.7 to the latest.

ocefpaf replied:
Thanks @lsetiawan! I will add a comment there to document this.

miguelcleon replied:
I didn't like the way the newer versions looked so I locked it into this older version. @lsetiawan what breaks for you? I haven't noticed any problems. It would be nice to be able to remove the pin but that might require using a different style app.

miguelcleon replied:
This is kind of arbitrary, I suppose. there is nothing wrong, per say with the newer versions, at least that I know of.

lsetiawan replied:
@miguelcleon and @ocefpaf I take back what I said earlier. The latest version does not pose any weirdness anymore, the style seemed to stay the same. I tried creating a fresh environment with an unspecified version of djangocms-style-admin and it seems to work just fine ver. [1.2.6.2](https://pypi.python.org/pypi/djangocms-admin-style).

miguelcleon replied:
@lsetiawan  Ok I misunderstood your comment. I'm taking another look at the latest version. What do you think?

miguelcleon replied:
@lsetiawan yeah, I really don't like the way the shortcuts from django-admin-shortcuts look with the newer djangocms-admin-style.

lsetiawan replied:
@miguelcleon @ocefpaf This is what I get:
![image](https://dl.dropboxusercontent.com/u/64841654/Screenshot%202017-01-11%2015.13.39.png)

It is still not looking good. Sorry for the confusion. Nothing is broken though!

ocefpaf replied:
Sorry. I am confused. Should we keep the pin to preserver the style (I will update the comment) or should I remove the pin?

lsetiawan replied:
Sorry I confused everyone. You should keep the pin to preserve the style because the latest version looked different.

ocefpaf replied:
No problem. I am updating the comment.

miguelcleon replied:
Great, thanks both!

miguelcleon commented:
@ocefpaf I was able to build a conda python environment using odm2adminenv.yml and got it to run odm2 admin on my dev server! :) One Issue I did have is django-recaptcha caused an error so I added it to the environment manually but then everything worked. 

`CondaValueError: Value error: invalid package specification: pip - django-recaptcha
`

ocefpaf commented:
> @ocefpaf I was able to build a conda python environment using odm2adminenv.yml and got it to run odm2 admin on my dev server! :) One Issue I did have is django-recaptcha caused an error so I added it to the environment manually but then everything worked.
Here is the error I received:
CondaValueError: Value error: invalid package specification: pip - django-recaptcha
For example it is running the site here http://dev-odm2admin.cuahsi.org/LCZO/LCZO/mapdata.html

I fixed a missing `:` in https://github.com/miguelcleon/ODM2-Admin/pull/90/commits/dba300b60c334109f0522da1a4c4b3dfa68bd08f. However, bare in mind that I plan to remove the `odm2adminenv.yml` file once `django-recaptcha` makes into the `conda-forge` channel, that file only serves as a workaround for a missing dependency.

That is because we should not keep two separate set of instructions (`conda` and `pip`, `requirements.txt` and `environment.yml`). I plan to merge both like I did in the `.travis.yml` so both `conda` and `pip` users can install `ODM2-Admin` from the `requirements.txt`

miguelcleon commented:
oh well that's a simple fix. Now it worked without error! I created fresh environment from the odm2adminenv.yml file and had no problems this time.

ocefpaf commented:
> oh well that's a simple fix. 

Yep. I was purposely lazy with that file because, as I mention in https://github.com/miguelcleon/ODM2-Admin/pull/90#issuecomment-272246180, I mean to remove it once this PR is ready.

ocefpaf commented on tests/test_dummy.py:
```diff
@@ -0,0 +1,12 @@
+
+
+def test_dummy():
+    assert always_true
```
@lsetiawan this is just an example on how to write tests with pytest. We must remove it once we write real tests.

I am not qualified to write tests for `odm2admin` (yet) because I am not familiar with its internals. But feel free to ping me anytime if you need help writing tests.

ocefpaf commented:
@miguelcleon his is ready for review now.

PS: I added the linter checks to the `allow_failures` to prevent that from being a blocker for easy merges.
But `odm2adim` is accumulating quite a few `pep8` infractions. Those degrade the code quality and make it hard for third party contributors to help with the project.

https://travis-ci.org/miguelcleon/ODM2-Admin/jobs/191720679

miguelcleon commented:
Great, thanks! I'll take a look. Thanks for pointing out the pep8 issues, these will be addressed.

miguelcleon commented:
seems good to me.

miguelcleon commented:
@ocefpaf should running `setup.py install` create the conda environment from the odm2adminenv.yml file?

ocefpaf commented:
> @ocefpaf should running setup.py install create the conda environment from the odm2adminenv.yml file?

No. `setup.py` does not work with `odm2adminenv.yml`. I guess that I should clarify a few things:

1. Packaging for PyPI, and then conda, is not ready yet! There is a lot to be done to ensure a working package and I will submit more PRs soon.
2. `odm2adminenv.yml` is conda specific and I will be recommending its removal in my next PR. (More details soon.)
3. [listing](https://github.com/miguelcleon/ODM2-Admin/blob/master/setup.py#L65) the [dependencies](https://github.com/miguelcleon/ODM2-Admin/blob/master/setup.py#L42-L44) in `setup.py` will try to fetch them. But your mileage will vary pending on the dependencies availability  on PyPI.

Sadly the case for  `odm2admin`, due to its dependency on `psycopg2`, is not an easy one. This problem will be worse if we add `geodjango`. Note that we do not have this issue with conda because we provide binaries for all the dependencies.

miguelcleon commented:
Ok, right. Let me know when there is more stuff to test out.
